rfcs/027-first-class-queries.md

```diff
@@ -0,0 +1,81 @@
+
+## Reference-level Explanation (aka Technical Details)
+
+### Existing functionality
+
+This feature is intended to be implemented entirely separately from existing storage of usage patterns, at least initially. Once we get the implementation in place, we may choose to start using it as the normalized source of truth for more usage information.
+
+This would interact with `DashboardQuery`, in that this would be a superset of that functionality. I suggest implementing this first, and transitioning existing `DashboardQuery` code to use this more generic functionality.
+
+Eventually, this could subsume the denormalized form `TableColumnUsage` for installs that have full data available. It is not likely we would want to deprecate `TableColumnUsage` to allow for different types of integrations. It's possible we would want to pre-generate `TableColumnUsage` from the underlying data, both for compatibility and read-time performance reasons.
+
+### databuilder
+
+We would create a new `class Query(GraphSerializable)` . It would contain these properties:
+
+- Time that the query was run (required)
+- At least one of required:
+    - Raw string representation of the query (for query types like SQL)
+    - Name of job run (for things like Python notebooks run)
+- Optional links to other data:
+    - User who ran the query
+    - Table(s)
+        - This relationship optionally includes info like:
+            - Relationship type: query from, join?
+                - For joins, include the join condition.
+    - Column(s) (if given, Tables may possibly be inferred from this)
+        - Optionally including info like WHERE clause conditions
+
+This would emit to neo4j in the shape you'd expect: a new node for queries, and relationships to existing stuff.
+
+Given that writing the SQL parsing logic is out-of-scope for this RFC, an `Extractor` is out-of-scope.
+
+## Drawbacks
+
+This is purely additive right now, so there are few downsides.
+
+Considering the future where this is used more heavily to support other features, however: it is possible that query volume would be extremely high, and thus storing in neo4j would be impractical. Pruning old queries would now change the semantics of certain queries: e.g. instead of popular tables being all-time popular, they would only be popular over the expiry period.
+
+## Alternatives
+
+The primary alternative is to continue relying on other systems to aggregate usage information, and ingesting into Amundsen aggregates of such information, rather than the source.
+
+## Prior art
+
+- Issue mentioned by jonhehir in [https://github.com/amundsen-io/amundsen/issues/547](https://github.com/amundsen-io/amundsen/issues/547)
+
+## Unresolved questions
+
+- Is there other prior art or considerations about this? Surely it's been considered before.
```

Comment:
I think having a table query relationship makes sense, however, I am not sure about using amundsen as the source of truth for all the queries which could easily store many non-sense queries(e.g typo etc).

1. SLA of query log emit from computation framework to this store is NA (typical is to hive table)
2. Non GC of the query nodes, are you going to retain PII data(query has all kinds PII) for the full deployment?

I think we could  have the query model, but if I am going to use it, I will only surface the most relevant / popular queries to the table (top 5, or top X)

Reply:
Yeah I think in our use cases, we would _probably_ lean toward only surfacing either queries that were run recently, or queries that are associated with dashboards - and we'd probably only want to store recent ones in Amundsen too.  I've run into a lot of situations where someone references a very old query, and the business logic has shifted enough since then that they end up making some incorrect assumptions